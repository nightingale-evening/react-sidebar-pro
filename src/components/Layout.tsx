import React from 'react';
import styled from 'styled-components';
import classnames from 'classnames';
import { useSidebar } from './sidebarContext';
import { useLayout } from './layoutContext';

export interface LayoutProps extends React.HTMLAttributes<HTMLDivElement> {
  rtl?: boolean;

  /**
   * set to ```true``` when Sidebar component is a direct descendent of Layout
   * @default ```false```
   */
  hasSidebar?: boolean;
}

interface StyledLayoutProps extends LayoutProps {
  fixedSidebar?: boolean;
}

const StyledLayout = styled.div<StyledLayoutProps>`
  min-height: 100%;
  overflow: auto;
  display: flex;
  flex-direction: column;
  position: relative;
  flex-grow: 1;
  ${({ rtl }) => (rtl ? 'direction: rtl' : '')};
  ${({ hasSidebar }) => (hasSidebar ? 'flex-direction: row' : '')};
  ${({ fixedSidebar }) => (fixedSidebar ? 'height: 100%' : '')};
`;

export const Layout: React.FC<LayoutProps> = ({ rtl = false, children, className, ...rest }) => {
  const { fixed: fixedSidebar } = useSidebar();
  const { rtl: layoutRtl, updateLayoutState } = useLayout();

  React.useEffect(() => {
    updateLayoutState({ rtl });
  }, [rtl, updateLayoutState]);

  return (
    <StyledLayout
      data-testid="layout-test-id"
      fixedSidebar={fixedSidebar}
      className={classnames('layout', className)}
      rtl={layoutRtl}
      {...rest}
    >
      {children}
    </StyledLayout>
  );
};
