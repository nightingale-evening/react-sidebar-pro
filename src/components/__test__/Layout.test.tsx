import React from 'react';
import { customRender, screen } from '../../utils/testUtils';
import { Layout } from '../Layout';

describe('Layout', () => {
  it('basic snapshot ', () => {
    const { container } = customRender(<Layout>Layout content</Layout>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should initialize Layout correctly', async () => {
    customRender(<Layout>Layout content</Layout>);
    const layoutElem = screen.getByText('Layout content');

    expect(layoutElem).toBeInTheDocument();
    expect(layoutElem).toHaveClass('layout');
    expect(layoutElem).toHaveStyle({
      display: 'flex',
      'flex-direction': 'column',
      'flex-grow': '1',
    });
  });

  it('should render flex-direction: row when hasSidebar is true', async () => {
    customRender(<Layout hasSidebar>Layout content</Layout>);
    const layoutElem = screen.getByTestId('layout-test-id');

    expect(layoutElem).toHaveStyle({
      'flex-direction': 'row',
    });
  });

  it('should render direction:rtl when rtl prop is true', async () => {
    customRender(<Layout rtl>Layout content</Layout>);
    const layoutElem = screen.getByTestId('layout-test-id');

    expect(layoutElem).toHaveStyle({
      direction: 'rtl',
    });
  });
});
