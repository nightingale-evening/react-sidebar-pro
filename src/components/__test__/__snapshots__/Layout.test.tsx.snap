// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout basic snapshot  1`] = `
<div
  class="sc-bdvvtL eMoNwh layout"
  data-testid="layout-test-id"
>
  Layout content
</div>
`;
